File: python/copass-management/src/copass_management/registrar.py
"""Transport-agnostic management-tool registrar.

Loads the management spec corpus, builds JSON-Schema validators for each
tool's input/output schema, wires every spec entry to its
``copass-core`` handler, and calls ``register(...)`` once per tool.

The transport (MCP SDK, backend tool resolver, plain function table) is
the caller's concern — only the optional ``adapters/mcp.py`` knows about
MCP.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any, Awaitable, Callable, Dict, List, Optional

from copass_core import CopassClient
from jsonschema import Draft202012Validator

from copass_management.specs import (
    ManagementSpec,
    load_management_specs,
)
from copass_management.tools import TOOL_HANDLERS


@dataclass(frozen=True)
class ToolContext:
    client: CopassClient
    sandbox_id: str


ToolHandler = Callable[[ToolContext, Dict[str, Any]], Awaitable[Any]]


@dataclass(frozen=True)
class ToolRegistration:
    name: str
    description: str
    input_schema: Dict[str, Any]
    output_schema: Dict[str, Any]
    input_validator: Draft202012Validator
    output_validator: Draft202012Validator
    handler: Callable[[Dict[str, Any]], Awaitable[Any]]
    spec: ManagementSpec


Register = Callable[[ToolRegistration], None]


@dataclass(frozen=True)
class RegistrarOptions:
    sandbox_id: str
    spec_dir: Optional[Path] = None
    validate_output: bool = False


def register_management_tools(
    register: Register,
    client: CopassClient,
    options: RegistrarOptions,
) -> List[ToolRegistration]:
    """Walk the spec corpus, wire handlers, and call ``register`` per tool.

    Returns the list of registrations in spec-name order.
    """
    corpus = load_management_specs(spec_dir=options.spec_dir)
    ctx = ToolContext(client=client, sandbox_id=options.sandbox_id)
    registrations: List[ToolRegistration] = []

    for name in sorted(corpus.specs):
        spec = corpus.specs[name]
        handler = TOOL_HANDLERS.get(name)
        if handler is None:
            raise RuntimeError(
                f"register_management_tools: no handler implementation for tool "
                f"{name!r}. Add one in copass_management/tools/."
            )

        input_validator = Draft202012Validator(spec.input_schema)
        output_validator = Draft202012Validator(spec.output_schema)

        async def wrapped(
            raw_input: Dict[str, Any] | None,
            *,
            _handler: ToolHandler = handler,
            _input_validator: Draft202012Validator = input_validator,
            _output_validator: Draft202012Validator = output_validator,
            _ctx: ToolContext = ctx,
        ) -> Any:
            payload = raw_input or {}
            _input_validator.validate(payload)
            result = await _handler(_ctx, payload)
            if options.validate_output:
                _output_validator.validate(result)
            return result

        registration = ToolRegistration(
            name=spec.name,
            description=spec.description,
            input_schema=spec.input_schema,
            output_schema=spec.output_schema,
            input_validator=input_validator,
            output_validator=output_validator,
            handler=wrapped,
            spec=spec,
        )
        register(registration)
        registrations.append(registration)

    return registrations


__all__ = [
    "ToolContext",
    "ToolHandler",
    "ToolRegistration",
    "Register",
    "RegistrarOptions",
    "register_management_tools",
]

File: typescript/packages/management/src/registrar.ts
import type { CopassClient } from '@copass/core';
import type { ZodTypeAny } from 'zod';

import { jsonSchemaToZod } from './json-schema-to-zod.js';
import {
  loadManagementSpecs,
  type LoadOptions,
  type ManagementSpec,
} from './specs.js';
import { TOOL_HANDLERS } from './tools/index.js';

export interface ToolContext {
  client: CopassClient;
  sandboxId: string;
}

export type ToolHandler = (
  ctx: ToolContext,
  input: Record<string, unknown>,
) => Promise<unknown>;

export interface ToolRegistration {
  name: string;
  description: string;
  /** Raw JSON Schema for the tool's input. */
  inputSchema: Record<string, unknown>;
  /** Raw JSON Schema for the tool's output. */
  outputSchema: Record<string, unknown>;
  /** Compiled Zod parser for the input. */
  inputZod: ZodTypeAny;
  /** Compiled Zod parser for the output. */
  outputZod: ZodTypeAny;
  /** Async handler that runs the call and returns the parsed result. */
  handler: (input: unknown) => Promise<unknown>;
  /** The full spec object as loaded from disk. */
  spec: ManagementSpec;
}

export type Register = (registration: ToolRegistration) => void;

export interface RegistrarOptions extends LoadOptions {
  /** Sandbox the registered tools target. Required — every Phase 1 read tool
   * is sandbox-scoped. */
  sandboxId: string;
  /**
   * When true, validate the handler's HTTP response against the tool's
   * `outputSchema` before returning. Defaults to `false` because handlers
   * already pass through `@copass/core`-typed responses; the conformance
   * test enforces parity at build time. Enabling this adds runtime cost.
   */
  validateOutput?: boolean;
}

/**
 * Transport-agnostic management-tool registrar.
 *
 * Loads the management spec corpus, builds Zod parsers for each tool's
 * input/output schema, wires every spec entry to its `@copass/core`
 * handler, and calls `register(...)` once per tool.
 *
 * The transport (MCP SDK, backend tool resolver, plain function table)
 * is the caller's concern.
 */
export function registerManagementTools(
  register: Register,
  client: CopassClient,
  options: RegistrarOptions,
): ToolRegistration[] {
  const corpus = loadManagementSpecs(options);
  const ctx: ToolContext = { client, sandboxId: options.sandboxId };
  const registrations: ToolRegistration[] = [];

  const names = Object.keys(corpus.specs).sort();
  for (const name of names) {
    const spec = corpus.specs[name];
    const handler = TOOL_HANDLERS[name];
    if (!handler) {
      throw new Error(
        `registerManagementTools: no handler implementation for tool "${name}". Add one in src/tools/.`,
      );
    }

    const inputZod = jsonSchemaToZod(spec.inputSchema);
    const outputZod = jsonSchemaToZod(spec.outputSchema);

    const wrapped = async (rawInput: unknown): Promise<unknown> => {
      const parsedInput = inputZod.parse(rawInput ?? {}) as Record<string, unknown>;
      const result = await handler(ctx, parsedInput);
      if (options.validateOutput) {
        return outputZod.parse(result);
      }
      return result;
    };

    const registration: ToolRegistration = {
      name: spec.name,
      description: spec.description,
      inputSchema: spec.inputSchema,
      outputSchema: spec.outputSchema,
      inputZod,
      outputZod,
      handler: wrapped,
      spec,
    };

    register(registration);
    registrations.push(registration);
  }

  return registrations;
}
